fix: prevent data url render

ojeytonwilliams opened the pull request:
This will stop any existing data urls from being rendered. Users trying to modify their profile image will be warned that they need to provide a link.

ojeytonwilliams committed:
fix: only render profile urls

ojeytonwilliams committed:
feat: warn user when submitting data url

raisedadead reviewed (commented):
Not directly related, but for every check that we add on the Client we should also have a check in the API (think strict typing) too.

ojeytonwilliams commented:
We can, but the objective here was just to make sure that we're not hosting anything we shouldn't.  If an inappropriate URL gets submitted the api will store it, but the client won't display it.

raisedadead commented:
> If an inappropriate URL gets submitted the api will store it, but the client won't display it.

Like I said, this is unrelated and pretty soon the "client" will not be the only client interacting with the API.

ojeytonwilliams commented:
> pretty soon the "client" will not be the only client interacting with the API

Ah, good point.  I might as well take a look, then.  Given that's the case, should we remove these urls from the db?

ojeytonwilliams committed:
fix: prevent dataurls being saved to db

raisedadead commented:
>  Given that's the case, should we remove these urls from the db?

We should be able to do "cleanups" on an ad-hoc basis. That is whenever we update the user object (or any document) for that matter in the DB, we should make it "clean" and schema compliant.

Could be it's own implementation though.

ShaunSHamilton commented on client/src/components/helpers/avatar-renderer.tsx:
```diff
@@ -26,9 +27,13 @@ function AvatarRenderer({
-    validationImage.src = picture;
-    validationImage.onload = onImageLoad;
-    validationImage.onerror = onImageError;
+    if (isURL(picture)) {
```
Do we not want to include some of the options like `{require_protocol: true}`

ojeytonwilliams replied:
It's not strictly necessary since any invalid url should result in an error when we `src` it.  We can, though. 

Which ones did you have in mind?

ShaunSHamilton replied:
As far as I can see, that is the only default one I would change. I forgot about our sourcing tests, though. So, it is fine either way.

ojeytonwilliams committed:
fix: tests and imports

Not sure why jest didn't like the es imports, but they aren't necessary
so I dropped them.

ojeytonwilliams commented:
> We should be able to do "cleanups" on an ad-hoc basis. 

I'm a little leery of db queries along the lines of "starts with data" since that sounds a bit slow, but I'm a little out of my depth when it comes to mongo.

ojeytonwilliams committed:
fix: check for url protocol

raisedadead commented:
No, what I meant is NOT doing sweeps. 

Instead, we normalize the whole user document during the update operations. That way we can do a form of eventual consistency, that is: 
- When a document is required in the real world it gets sanitized and presented. 
- When a document is updated in the real world it gets sanitized + made compliant with the latest schema/default values and pushed up.

ojeytonwilliams commented:
Oh, sorry, I thought we were talking about something else. Don't we want to audit these in case we forget to screen them out of a new client?

In general I like the approach you proposed, but I thought we needed to handle this specific case differently.

raisedadead commented:
> In general I like the approach you proposed, but I thought we needed to handle this specific case differently.

Yes - absolutely, what I am proposing needs careful planing and changes to the API.

ojeytonwilliams commented:
Okay, I'm happy with this, then.  I've added a TODO for the planning bit, but we should hop on a call at some point.

ShaunSHamilton reviewed (approved):
I tried quite hard to break this, because I was suspicious about the difference between the `onChange` for `FormControl` returning vs not returning.

So, this LGTM 👍 

Personally, I hope we can remove this, and opt for the server-side validation.
